File: starrysky/src/main/java/com/lzx/starrysky/common/BaseKey.java
package com.lzx.starrysky.common;

import java.util.ArrayDeque;
import java.util.Queue;

abstract public class BaseKey<T> {

    private final Queue<T> keyPool = new ArrayDeque<>();

    public T get() {
        T result = keyPool.poll();
        if (result == null) {
            result = create();
        }
        return result;
    }

    public void offer(T key) {
        keyPool.offer(key);
    }

    public abstract T create();
}

File: starrysky/src/main/java/com/lzx/starrysky/common/BaseKey.kt
package com.lzx.starrysky.common

import java.util.ArrayDeque
import java.util.Queue

abstract class BaseKey<T> {

    private val keyPool = ArrayDeque<T>()

    fun get(): T? {
        var result: T? = keyPool.poll()
        if (result == null) {
            result = create()
        }
        return result
    }

    fun offer(key: T) {
        keyPool.offer(key)
    }

    abstract fun create(): T
}
